### PR TITLE
Dnadales/538 create stylish haskell configuration and instructions

### LINKED ISSUE #538: Create a stylish Haskell configuration and instructions on how to use it
In order to have consistent formatting of language extensions declarations and module imports we should use stylish Haskell. We can use as inspiration the [configuration used in `cardano-sl`](https://github.com/input-output-hk/cardano-sl/blob/develop/.stylish-haskell.yaml).

The idea is to:

- create a stylish Haskell configuration that the team agrees with
- add instructions on how to use it

### PR DESCRIPTION
**Pull request by dnadales**
Closes #538

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+indent_style = space
+indent_size = 2
+end_of_line = lf
+charset = utf-8
+trim_trailing_whitespace = true
+insert_final_newline = true
+
+[*.hs]
+indent_size = 2
+max_line_length = 80
```
Comment (dnadales) [inline .editorconfig]: Do we want to restrict this to 80, or does 100 makes more sense?

File: .editorconfig [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+[*]
+indent_style = space
+indent_size = 2
+end_of_line = lf
+charset = utf-8
+trim_trailing_whitespace = true
+insert_final_newline = true
+
+[*.hs]
+indent_size = 2
```
Comment (dnadales) [inline .editorconfig]: I guess we're already using an indent size of two, so this shouldn't be controversial.

File: .stylish-haskell.yaml [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+      #                          init, last, length, scanl, scanr, take, drop,
+      #                          sort, nub)
+      #
+      # - new_line: Import list starts always on new line.
+      #
+      #   > import qualified Data.List      as List
+      #   >     (concat, foldl, foldr, head, init, last, length)
+      #
+      # Default: after_alias
+      list_align: with_module_name
```
Comment (dnadales) [inline .stylish-haskell.yaml]: This option was recently added by Ru, so we need a recent version of `stylish-haskell`

File: .editorconfig (hunk 1 above)
Comment (erikd) [reply .editorconfig]: I would vote for 100.

**Comment by erikd** [comment]
There are (IMO) a number of issues with what `stylish-haskell` does with this setup.

1. Aligns closing `#-}` of `LANGUAGE` pragmas. This is suboptimal, because as a pragmas are added or removed this can result in non-minimal diffs.
1. It vertically aligns the `::` of all fields of product types. Again, as types are added/removed/renamed, non-minimal diffs are generated.
1. It vertically aligns the 'as` component of `import qualified X.Y.Z as A`. The problem is the otential for non-minimal diffs.
1. It vertically aligns the `->` of pattern matches. The problem is the otential for non-minimal diffs.
1. Import lists are take up significantly more vertical space than they should.

There may be other issues I missed.

**Comment by erikd** [comment]
The following patch to `.stylish-haskell.yaml`:
```
diff --git a/.stylish-haskell.yaml b/.stylish-haskell.yaml
index 328811c..bc7e8e0 100644
--- a/.stylish-haskell.yaml
+++ b/.stylish-haskell.yaml
@@ -19,9 +19,9 @@ steps:
   # and only applies to statements where each element occupies a single
   # line. All default to true.
   - simple_align:
-      cases: true
+      cases: false
       top_level_patterns: true
-      records: true
+      records: false
 
   # Import cleanup
   - imports:
@@ -197,7 +197,7 @@ steps:
       #   between actual import and closing bracket.
       #
       # Default: true
-      align: true
+      align: false
 
       # stylish-haskell can detect redundancy of some language pragmas. If this
       # is set to true, it will remove those redundant pragmas. Default: true.
```
The first two disable vertical alignment of pattern matches and record field types. The second disables alignment of the closing `#-}` of language pragmas.

**Comment by dnadales** [comment]
> There are (IMO) a number of issues with what `stylish-haskell` does with this setup.
> 
> 1. Aligns closing `#-}` of `LANGUAGE` pragmas. This is suboptimal, because as a pragmas are added or removed this can result in non-minimal diffs.
> 2. It vertically aligns the `::` of all fields of product types. Again, as types are added/removed/renamed, non-minimal diffs are generated.
> 3. It vertically aligns the 'as`component of`import qualified X.Y.Z as A`. The problem is the otential for non-minimal diffs.
> 4. It vertically aligns the `->` of pattern matches. The problem is the potential for non-minimal diffs.
> 5. It vertically aligns the `=` of functions that use pattern matching. The problem is the potential for non-minimal diffs.
> 
Yeah, I think the non-minimal diffs are a good point. We should maybe privilege this over a more visually pleasing result.


> There may be other issues I missed, but I am pleased to see that import lists don't take up as much vertical space.
> 
:+1: Do you have any suggestions for tweaks in the SH config to improve this?

> In addition some of the files in this repo (eg `byron/ledger/executable-spec/src/Ledger/Core.hs`) cannot be parsed by `stylish-haskell` because the later uses `haskell-src-exts` to parse Haskell code and that package does not support some more recent features of Haskell.

Yeah, but I guess there's no much we can do there.

**Comment by dnadales** [comment]
Thanks for your comments @erikd I incorporated your suggestions.

**Comment by erikd** [comment]
Yeah, I can live with what we have here.

**Comment by dnadales** [comment]
> Yeah, I can live with what we have here.

After normalizing that comment accounting for the Erik scale we get a very enthusiast endorsement for this PR :P

File: .stylish-haskell.yaml (hunk 3 above)
Comment (nc6) [reply .stylish-haskell.yaml]: Is this version actually on hackage/nixpkgs? The one thing I'm completely against is having to rely on a custom version of stylish-haskell, as is required in `cardano-sl`, for example.

Comment (dnadales) [reply .stylish-haskell.yaml]: Yes, we merged that into `stylish-haskell` a couple of weeks ago.

**Comment by dnadales** [comment]
Yes, it is on `hackage` http://hackage.haskell.org/package/stylish-haskell

https://github.com/jaspervdj/stylish-haskell/compare/v0.9.2.2...master

File: .dir-locals.el [hunk 4]
```diff
@@ -0,0 +1,8 @@
+;;; Directory Local Variables
+;;; For more information see (info "(emacs) Directory Variables")
+
+((nil
+  (haskell-mode-stylish-haskell-path . "stylish-haskell")
+  )
+ (haskell-mode
+  (haskell-stylish-on-save . t)))
```
Comment (mgudemann) [inline .dir-locals.el]: do we really want an editor specific file ?

File: .stylish-haskell.yaml [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+  - trailing_whitespace: {}
+
+  # Squash multiple spaces between the left and right hand sides of some
+  # elements into single spaces. Basically, this undoes the effect of
+  # simple_align but is a bit less conservative.
+  # - squash: {}
+
+# A common setting is the number of columns (parts of) code will be wrapped
+# to. Different steps take this into account. Default: 80.
+columns: 100
```
Comment (mgudemann) [inline .stylish-haskell.yaml]: is there a reason for deviating from 80? Obviously any choice of number is a little arbitrary ...

Comment (dnadales) [reply .stylish-haskell.yaml]: 100 gives us a little more (25%) horizontal room. But I have no strong preference for one or the other.

File: .dir-locals.el (hunk 4 above)
Comment (dnadales) [reply .dir-locals.el]: That's a good question. By having this here we make sure that Emacs users don't have to setup anything extra (and if they don't use `stylish-haskell` for other projects, they don't have to set this up locally).

Alternatively I could put this in the README, but then everyone who uses Emacs would have to copy and paste this.

I don't know what's more sensible ...

File: .stylish-haskell.yaml (hunk 5 above)
Comment (mgudemann) [reply .stylish-haskell.yaml]: ok, no problem, 100 is fine with me

File: .dir-locals.el (hunk 4 above)
Comment (mgudemann) [reply .dir-locals.el]: People might have their own `.dir-locals.el` file and get a conflict with this. The might also use a different way to setup stuff. But in reality it probably doesn't matter. Let's do it this way and see whether someone complains.

Comment (dnadales) [reply .dir-locals.el]: Just replying here as well for the record: the reasons I went for including `.dir-locals.el` into our git repo are:

-  if people need a custom `.dir-locals.el` to edit files in this project then I think there's a high chance that other developers might want to use those settings too.
- by having a versioned `.dir-locals.el` everybody that uses Emacs don't need additional setup.

**Pull request merged**